x-pack/plugins/security_solution/public/management/services/exceptions_list/exceptions_list_api_client.ts

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CreateExceptionListItemSchema,
+  CreateExceptionListSchema,
+  ExceptionListItemSchema,
+  ExceptionListSummarySchema,
+  FoundExceptionListItemSchema,
+  ListId,
+  UpdateExceptionListItemSchema,
+} from '@kbn/securitysolution-io-ts-list-types';
+import { EXCEPTION_LIST_ITEM_URL, EXCEPTION_LIST_URL } from '@kbn/securitysolution-list-constants';
+import { HttpStart } from 'kibana/public';
+import { MANAGEMENT_DEFAULT_PAGE, MANAGEMENT_DEFAULT_PAGE_SIZE } from '../../common/constants';
+
+/**
+ * A generic class to be used for each artifact type.
+ * It follow the Singleton pattern.
+ * Please, use the getInstance method instead of creating a new instance when using this implementation.
+ */
+export class ExceptionsListApiClient {
+  private static instance: Map<ListId, ExceptionsListApiClient> = new Map();
+  private static wasListCreated: Map<ListId, Promise<void>> = new Map();
+  private ensureListExists: Promise<void>;
+
+  constructor(
+    private readonly http: HttpStart,
+    private readonly listId: ListId,
+    private readonly listDefinition: CreateExceptionListSchema
+  ) {
+    this.ensureListExists = this.createExceptionList();
+  }
+
+  /**
+   * PrivateStatic method that creates the list and don't throw if list already exists.
+   * This method is being used when initializing an instance only once.
+   */
+  private async createExceptionList(): Promise<void> {
+    if (ExceptionsListApiClient.wasListCreated.has(this.listId)) {
+      return ExceptionsListApiClient.wasListCreated.get(this.listId);
+    }
+    ExceptionsListApiClient.wasListCreated.set(
+      this.listId,
+      new Promise<void>(async (resolve, reject) => {
+        try {
+          await this.http.post<ExceptionListItemSchema>(EXCEPTION_LIST_URL, {
+            body: JSON.stringify({ ...this.listDefinition, list_id: this.listId }),
+          });
+
+          resolve();
+        } catch (err) {
+          // Ignore 409 errors. List already created
+          if (err.response?.status !== 409) {
+            reject(err);
```

Comment:
If it failed, should we keep the rejected promise in `wasListCreated`? I would say clear it out because we know the list was not actually created.

Reply:
It is cleared [here](https://github.com/elastic/kibana/pull/124245/commits/b679c0305c9b7da0b405e767104887bac3eae5c3#diff-1421b2a7a753cb745574cd6862bea79e1eb205abcc45198780c6e8b36ef1eb75R68)

Reply:
also, should you `return;` here? (since this is an async function?

(actually - what do Promise's do when the the execution function is async? not sure the behaviour here)